KEP-5313: Placement Decision API for multicluster scheduling

mikeshng opened the pull request:
- One-line PR description:  Add a new KEP to introduce the Placement Decision API for multicluster scheduling

<!-- link to the k/enhancements issue -->
- Issue link: https://github.com/kubernetes/enhancements/issues/5313

<!-- other comments or additional information -->
- Other comments:

/sig multicluster

mikeshng commented:
/assign @deads2k @RainbowMango @zhiying-lin

CC  @corentone @elgnay @haoqing0110 @jnpacker  @qiujian16 @ryanzhang-oss

haoqing0110 commented on keps/sig-multicluster/5313-placement-decision-api/README.md:
```diff
@@ -0,0 +1,412 @@
+* All slices for one `Placement` MUST carry the same
+  `multicluster.x-k8s.io/placement=<placement-name>` label so consumers can List with a label selector.
+* The scheduler that owns the slice MUST label it with
+  `multicluster.x-k8s.io/scheduler=<scheduler-name>` for observability.
```
One question, does the API need to specify the order in which clusters are arranged? as well as the order of PlacementDecision.

haoqing0110 replied:
Maybe an optional label like `multicluster.x-k8s.io/decision-index=<index>`, if the clusters in the decision are ordered, the scheduler should add this label, especially when the result exceeds 100 clusters.

ryanzhang-oss replied:
+1, order might be userful

mikeshng replied:
Updated as suggested by CNCF Ambassador Qing Hao ;)

zhiying-lin commented on keps/sig-multicluster/5313-placement-decision-api/README.md:
```diff
@@ -0,0 +1,412 @@
+  Decisions []ClusterDecision `json:"decisions"`
+}
+
+// Label that links all slices to their originating Placement request.
+const PlacementLabel = "multicluster.x-k8s.io/placement"
+
+// Label that link all slices to which scheduler produced the slices.
+const SchedulerLabel = "multicluster.x-k8s.io/scheduler"
```
should we make it clear that these two labels are immutable during the lifecycle?

mikeshng replied:
Yes, I've added a new lifecycle section that mention this.

zhiying-lin commented on keps/sig-multicluster/5313-placement-decision-api/README.md:
```diff
@@ -0,0 +1,412 @@
+
+* Max size: 100 ClusterDecision entries per slice keeps object well below etcd limit.
+
+* Validation: A webhook may verify that every clusterName exists
```
the placement label value should not be empty?
the placement & scheduler label value should be immutable?

mikeshng replied:
Done as suggested. Thanks!

zhiying-lin commented on keps/sig-multicluster/5313-placement-decision-api/README.md:
````diff
@@ -0,0 +1,412 @@
+
+```
+// PlacementDecision publishes the selected clusters for one Placement.
+type PlacementDecision struct {
````
do we need to define what's the behavior after placementDecision is deleted? for example, all the resources will be deleted from the selected clusters. So that, the consumer will have a right expectation on this.

mikeshng replied:
Yes! I've added a Lifecycle section. Thank you for the great feedback!

corentone replied:
I'd say it's syncer/controller dependent. Also, there are a bunch of scenario where it may not be possible: loss of privileges, cluster disconnected, clusterprofile gone.

iholder101 commented:
/cc @awels
FYI

corentone commented on keps/sig-multicluster/5313-placement-decision-api/README.md:
```diff
@@ -0,0 +1,429 @@
+* Define a namespaced scope, minimalistic, data only `PlacementDecision` API that lists selected clusters.
+* Support continuous rescheduling: decision list may be updated.
+* Guarantee that every `clusterName` entry matches a `ClusterProfile.metadata.name` in the same inventory.
+* Guarantee that every `clusterName` entry is in the same namespace as `PlacementDecision.metadata.namespace`.
```
I think this one is wrong, the inventory will likely be in a separate namespace from the placements/workloads.

corentone commented on keps/sig-multicluster/5313-placement-decision-api/README.md:
```diff
@@ -0,0 +1,429 @@
+* Support continuous rescheduling: decision list may be updated.
+* Guarantee that every `clusterName` entry matches a `ClusterProfile.metadata.name` in the same inventory.
+* Guarantee that every `clusterName` entry is in the same namespace as `PlacementDecision.metadata.namespace`.
+* Provide label conventions so consumers can retrieve all slices of one placement.
```
I wonder if we actually need this added complexity yet? How many clusters would a placement REALLY need to target? If we get more than a 100, maybe we're not using the right grouping/abstraction?

mikeshng replied:
There are fleets running across thousands of clusters, especially in edge and telco cases. While not everyday case, we have customers do push configs to all of those clusters at once.
The core issue with dumping large lists into a single object is that every change results in a large write. K8s API authors already designed APIs like EndpointSlice to handle watch/write churn and workaround etcd limits. So it makes sense to follow established conventions. Expecting users to manually shard their Placement objects to avoid etcd limits or expensive writes feels like a step backward in API design. CC @deads2k

corentone commented on keps/sig-multicluster/5313-placement-decision-api/README.md:
```diff
@@ -0,0 +1,429 @@
+
+* Decouples schedulers from consumers swap either side without rewriting the other.
+* Aligns with the SIG-Multicluster `ClusterProfile` inventory:
+  the decision's cluster keys are guaranteed to exist.
```
I agree with "Aligns with the SIG-Multicluster `ClusterProfile` inventory", but the l.95 is not true. It's a distributed system, the cluster may go and leave placements mentioning it. I'd just remove this justification.
```suggestion
```

mikeshng replied:
Updated as suggested.

corentone commented on keps/sig-multicluster/5313-placement-decision-api/README.md:
```diff
@@ -0,0 +1,429 @@
+* Aligns with the SIG-Multicluster `ClusterProfile` inventory:
+  the decision's cluster keys are guaranteed to exist.
+* Simplifies RBAC: one resource and label schema to secure.
+* Enables GPU aware AI scheduling, cost based placement, and rapid failover.
```
I don't think that's really true.
There is a fundamental problem with scheduling for HW, which is that the cluster's scheduler has the final say. The hub's placement scheduler is likely to have an incomplete view.

Also, PlacementDecision doesn't do anything about GPU/HW etc. That would be up to scheduler implementations :)

mikeshng replied:
Removed as suggested.

corentone commented on keps/sig-multicluster/5313-placement-decision-api/README.md:
```diff
@@ -0,0 +1,429 @@
+forcing downstream tools such as GitOps engine, workload orchestrator, progressive rollout controller,
+or AI/ML pipeline having to understand a scheduler specific API.
+
+This KEP introduces a vendor neutral `PlacementDecision` API that standardizes
```
Let's actually call it `Placement`? That gives us a chance to align the Spec later.

mikeshng replied:
We are using the name PlacementDecision on purpose to show it is only responsible for the scheduler's data only answer to "which clusters should be used?" It's different from any future standard Placement API (or current vendor specific Placement/Scheduling APIs) that defines the request/spec driving that decision.

corentone commented on keps/sig-multicluster/5313-placement-decision-api/README.md:
```diff
@@ -0,0 +1,429 @@
+
+A typical multicluster setup:
+
+1. Scheduler: examines the fleet (`ClusterProfile` objects) and decides *where* a workload should land.
```
```suggestion
1. Scheduler: examines the fleet (`ClusterProfile` objects), and other signals/metrics and decides *where* a workload should land.
```

mikeshng replied:
Updated as suggested. Thanks for the good catch!

corentone commented on keps/sig-multicluster/5313-placement-decision-api/README.md:
```diff
@@ -0,0 +1,429 @@
+so each consumer might need to learn different APIs which slow down integration,
+locking users to specific vendors, and complicate RBAC/validation work.
+
+A standardize unversial `PlacementDecision` API:
```
```suggestion
A standardized `PlacementDecision` API:
```

corentone commented on keps/sig-multicluster/5313-placement-decision-api/README.md:
```diff
@@ -0,0 +1,429 @@
+
+### Non-Goals
+
+* Describing how a scheduler made its choice (Placement API spec).
```
one of our plan for MCO is to publish events on the Why; so while the placement itself shouldn't care, the end user may care (for debugging purposes)

mikeshng replied:
Added the `Reason` field for end user.

corentone commented on keps/sig-multicluster/5313-placement-decision-api/README.md:
```diff
@@ -0,0 +1,429 @@
+  metav1.TypeMeta   `json:",inline"`
+  metav1.ObjectMeta `json:"metadata,omitempty"`
+
+  // Up to 100 ClusterDecisions per object (slice) to stay well below the etcd limit.
```
Should we remove this limitation of 100 clusters? that way we could avoid the whole idea of having to compose multiple placementdecision CRs together.

ryanzhang-oss replied:
I am not sure if we can get out of multiple placementdecision CRs given the ETCD limit.

mikeshng replied:
I agree with Ryan. See https://github.com/kubernetes/enhancements/pull/5314#discussion_r2098396579

corentone commented on keps/sig-multicluster/5313-placement-decision-api/README.md:
```diff
@@ -0,0 +1,429 @@
+const SchedulerLabel = "multicluster.x-k8s.io/scheduler"
+
+type ClusterDecision struct {
+  // MUST match a ClusterProfile.metadata.name in the same namespace.
```
Clusterprofile will most of the time be in a separate namespace I think.
So this part is not true and we should ask for both name and namespace.

corentone reviewed (commented):
Trying to simplify it a bit.

At the same time, will try to suggest sharing our MCO one as the placement.

qiujian16 commented on keps/sig-multicluster/5313-placement-decision-api/README.md:
```diff
@@ -0,0 +1,429 @@
+  namespace: argocd
+  labels:
+	  multicluster.x-k8s.io/placement: app-placement
+	  multicluster.x-k8s.io/scheduler: multicluster-placement-controller
```
why this is a label, but not a field?

qiujian16 commented on keps/sig-multicluster/5313-placement-decision-api/README.md:
```diff
@@ -0,0 +1,429 @@
+  `multicluster.x-k8s.io/scheduler=<scheduler-name>` for observability.
+
+### Lifecycle
+- **Create**: The scheduler creates the slice with the list of clusters in the decisions,
```
we might need some ownership explanation, e.g. scheduler is responsible for the api, consumers should not update but only read.

mikeshng replied:
Absolutely! Added a "Ownership" section as suggested.

qiujian16 commented on keps/sig-multicluster/5313-placement-decision-api/README.md:
```diff
@@ -0,0 +1,429 @@
+	  multicluster.x-k8s.io/placement: app-placement
+	  multicluster.x-k8s.io/scheduler: multicluster-placement-controller
+decisions:
+- clusterName: cluster1
```
will the decisions be changed by the scheduler too frequent? How to avoid that?

ryanzhang-oss replied:
another thought, should the decision be immutable?  i.e. The scheduler MUST produce another decision if the decision has changed.

mikeshng replied:
Added a paragraph on how to reduce frequent update churn. Thanks!

RainbowMango commented on keps/sig-multicluster/5313-placement-decision-api/README.md:
```diff
@@ -0,0 +1,412 @@
+Any scheduler can emit the object and any consumer can watch it.
+Neither side needs to know about the other, enabling true plug and play in the multicluster stack.
+
+![PlacementDecision Flow Diagram](./placementdecision.jpg)
```
Just out of curiosity, what's the mermaid thing? :)

qiujian16 replied:
https://mermaid.js.org/

RainbowMango commented on keps/sig-multicluster/5313-placement-decision-api/README.md:
```diff
@@ -0,0 +1,429 @@
+* Decouples schedulers from consumers swap either side without rewriting the other.
+* Aligns with the SIG-Multicluster `ClusterProfile` inventory:
+  the decision's cluster keys are guaranteed to exist.
+* Simplifies RBAC: one resource and label schema to secure.
```
Can you elaborate on why the label schema is secure? 
Compared to having a reference to `Placement` on `spec` might be more straightforward.

RainbowMango commented on keps/sig-multicluster/5313-placement-decision-api/README.md:
```diff
@@ -0,0 +1,429 @@
+
+* Describing how a scheduler made its choice (Placement API spec).
+* Describing how consumers access selected clusters.
+* Embedding orchestration logic or consumer feedback in `PlacementDecision`.
```
I remember Liqian asked the question at the last community meeting, 'How do we know the status, if the decision has been consumed?'.

Even though we haven't seriously talked about the Placement API, the Placement API should reference which workload goes where, and the status. If no status on `PlacementDecision`, where to get the status?

mikeshng replied:
PlacementDecision intentionally omits any scheduling/orchestration spec. Added a "Consumer Feedback" section to clarify that feedback is out of scope for this resource and should be handled by a separate mechanism.

RainbowMango commented on keps/sig-multicluster/5313-placement-decision-api/README.md:
```diff
@@ -0,0 +1,429 @@
+  metav1.ObjectMeta `json:"metadata,omitempty"`
+
+  // Up to 100 ClusterDecisions per object (slice) to stay well below the etcd limit.
+  Decisions []ClusterDecision `json:"decisions"`
```
So the `Decision` is located directly under the struct, no spec and status? It looks like the Secret or Config, not like a CRD.

mikeshng replied:
After some discussions, there are three potential locations (root vs status vs spec) for the cluster list. The root level (.decisions) location is preferred because it represents scheduler/placement-controller output (not user intent), allows labels and data to be written atomically in a single call, and follows the core EndpointSlice API design. See the details and discussions (even the resolved ones) in: https://docs.google.com/document/d/1seK6W_TgSDinogXqEm8bOgFCuKqJ9_qkZdodfkSheUY/edit?disco=AAABjJX9gjI&tab=t.0

ryanzhang-oss replied:
does CRD have to have spec/status? If so, where should the decision be?

mikeshng commented on keps/sig-multicluster/5313-placement-decision-api/README.md:
```diff
@@ -0,0 +1,429 @@
+* Describing how a scheduler made its choice (Placement API spec).
+* Describing how consumers access selected clusters.
+* Embedding orchestration logic or consumer feedback in `PlacementDecision`.
+* Replace/overlap Work API which is responsible for actually applying the workload.
```
Updated as suggested. Thanks!

mikeshng commented on keps/sig-multicluster/5313-placement-decision-api/README.md:
```diff
@@ -0,0 +1,412 @@
+* Guarantee that every `clusterName` entry matches a `ClusterProfile.metadata.name` in the same inventory.
+* Guarantee that every `clusterName` entry is in the same namespace as `PlacementDecision.metadata.namespace`.
+* Provide label conventions so consumers can retrieve all slices of one placement.
+* Leave room for schedulers implementations.
```
Yes! Added a point for the general approach for consumers.

mikeshng commented on keps/sig-multicluster/5313-placement-decision-api/README.md:
```diff
@@ -0,0 +1,491 @@
+
+  // Name of the Placement that this decision answers.
+  // Must equal the value of the label multicluster.x-k8s.io/placement.
+  PlacementName string `json:"placementName"`
```
Done! Thanks!

mikeshng commented on keps/sig-multicluster/5313-placement-decision-api/README.md:
```diff
@@ -0,0 +1,491 @@
+  PlacementName string `json:"placementName"`
+
+  // Up to 100 ClusterDecisions per object (slice) to stay well below the etcd limit.
+  Decisions []ClusterDecision `json:"decisions"`
```
Right this is perfect! I've added some more validations to other fields as well.

mikeshng commented on keps/sig-multicluster/5313-placement-decision-api/README.md:
```diff
@@ -0,0 +1,491 @@
+  ClusterNamespace string `json:"clusterNamespace"`
+
+  // Name of the target ClusterProfile.
+  ClusterName string `json:"clusterName"`
```
Yes, you are right! Followed the convention of other CRDs and created the clusterProfileRef struct to hold the namespacename.

mikeshng commented:
Closed threads I believe are resolved. Feel free to reopen or comment if there's more to discuss. Thanks!

ryanzhang-oss commented on keps/sig-multicluster/5313-placement-decision-api/README.md:
```diff
@@ -0,0 +1,537 @@
+  The scheduler may choose to populate the reason for each decision for consumers/end-users
+  (ie, for debugging purposes).
+
+- **Update / Reschedule**: The scheduler may add or remove clusters in  decisions at any time.
```
same comment as above, we need to make sure that the consumer knows when and if there is change in the decision.  I am not 100% sure that generation field works without spec.

mikeshng replied:
PlacementDecision changes should still be visible to consumers. Every update bumps the `resourceVersion` so standard informers fire an Update event each time the object changes and letting consumer controllers reconcile (or just diff the old and new decisions lists with a custom UpdateFunc in controller code). It should work the same way as how controllers handle `ConfigMap` with no spec/status updates.

ryanzhang-oss commented on keps/sig-multicluster/5313-placement-decision-api/README.md:
```diff
@@ -0,0 +1,537 @@
+
+#### Story 2: Progressive rollout
+
+* A progressive rollout controller begins with some canary clusters.  
```
I wonder when is the schedule decision made? Is it made before the starting of the rollout or on the fly? The system I've used mostly have the decision made before the rollout while the rollout component controls the progress.

mikeshng commented on keps/sig-multicluster/5313-placement-decision-api/README.md:
```diff
@@ -0,0 +1,537 @@
+}
+
+// ClusterProfileRef references a ClusterProfile by namespace and name.
+type ClusterProfileRef struct {
+	// Namespace of the referenced ClusterProfile.
+	// +kubebuilder:validation:MinLength=1
+	Namespace string `json:"namespace"`
+
+	// Name of the referenced ClusterProfile.
+	// +kubebuilder:validation:MinLength=1
+	Name string `json:"name"`
+}
```
Great question. I actually tried using the usual `NamespacedName` import, but since it doesn't have JSON tags so it's not compatible with CRD schemas. See: https://github.com/kubernetes/apimachinery/blob/master/pkg/types/namespacedname.go#L27-L30

mikeshng commented on keps/sig-multicluster/5313-placement-decision-api/README.md:
```diff
@@ -0,0 +1,537 @@
+// ClusterDecision references a target ClusterProfile for placement.
+type ClusterDecision struct {
+  // Reference to the target ClusterProfile.
+	ClusterProfileRef ClusterProfileRef `json:"clusterProfileRef"
```
Great catch. Updated as suggested. Thanks!

mikeshng commented:
> This looks good to me. There are a few typos etc. but we can take care of that later (I’ll follow up). I take it we’ll revisit the graduation criteria, test plans etc. after the initial merge, is that right?

Right, thanks @skitt !

Just pushed the update-toc check fix.

mikeshng commented on keps/sig-multicluster/5313-placement-decision-api/README.md:
```diff
@@ -0,0 +1,548 @@
+  The scheduler may choose to populate the reason for each decision for consumers/end-users
+  (ie, for debugging purposes).
+
+- **Update / Reschedule**: The scheduler may add or remove clusters in decisions at any time.
```
Added actions and examples on change.

mikeshng commented on keps/sig-multicluster/5313-placement-decision-api/README.md:
```diff
@@ -0,0 +1,548 @@
+
+### Terminology
+
+- **Placement**: A scheduler request that asks "where should this workload run?".
```
Updated the KEP to acknowledge this limitation.
While consumers may still need scheduler specific knowledge for complex scenarios, this API still provides value by standardizing the "where" (cluster list) output, enabling basic workload distribution and reducing integration complexity even without full placement request standardization.
WDYT Jeremy?

ryanzhang-oss replied:
Second to Mike. The value of this API is that it provides a single interface for any projects that have a multi-cluster component to out source the scheduling part.  For example, Argo's applicationSet can take a cluster generator, Multi-kueue now supports [external scheduling](https://kueue.sigs.k8s.io/docs/concepts/multikueue/#external-custom-implementation). With the schedulingDecision API,  we can now implement a common controller that can allow those project to tap into the scheduling capabilities that cluster managers projects (OCM/KubeFleet/Karma.. etc) provide.  I think there is clear value in it alone.

lauralorenz commented:
Triage notes:
* Conversation right now is about the origin position of having a placement decision API without a placement request API.
* For now added it as a non-goal -- is that enough?
* If not: at least looking for more motivation on the limits and scope of the placement decision API by itself without the placement request part
* Want to determine if this KEP can continue forward with the "one half" of the situation (response part without request)

skitt commented on keps/sig-multicluster/5313-placement-decision-api/README.md:
```diff
@@ -0,0 +1,559 @@
+    multicluster.x-k8s.io/placement: app-placement
+schedulerName: multicluster-placement-controller
+placementRef:
+  apiGroup: multicluster.x-k8s.io
+  kind: Placement
```
How would a non-scheduler-specific or aware consumer use this without knowledge of the `Placement` object?

mikeshng replied:
Great point. Removed PlacementRef as discussed in SIG-MC meeting.

mikeshng commented:
As discussed in the SIG-MC meeting, the KEP now removes `PlacementRef` entirely and no longer references a Placement, so PlacementDecision stands alone as data. There is a newly added a "Consumer Discovery and Usage" section that explains how a consumer can use label to correlate slices and provide an alternative deterministic naming approach as well. The KEP now included a short Argo CD example to show how a real consumer would find and use the decisions.

mikeshng commented:
Based on SIG-MC meeting feedback, the latest push highlights:
- Updated flow diagram to include the "left" side: initiator, placement and workload to clarify inputs.
- Added 2 concrete usage flows showing how PlacementDecisions are actually written and consumed.
- Added a section explaining why define PlacementDecision before a Placement API.

ryanzhang-oss commented on keps/sig-multicluster/5313-placement-decision-api/README.md:
```diff
@@ -0,0 +1,620 @@
+5. Scheduler/PlacementController writes one or more `PlacementDecision` slice objects, all correlated via the same `multicluster.x-k8s.io/decision-key`.
+6. A new controller watches/gets those `PlacementDecision` objects and finds the corresponding kueue `Workload` by placement key.
+7. The controller patches the `.status.nominatedClusterNames` field of the `Workload` with the list of clusters from the `PlacementDecision`.
+8. The MultiKueue controller then copies the `Workload` to the nominated clusters and waits for the `Workload` to be admitted by the `ClusterQueue` on the worker clusters.
```
yes, the only multi-cluster controller that is needed is a multi-cluster scheduler.  MultiKueue has its own dispatch mechanism.

ryanzhang-oss commented:
> From the SIG-MC meeting: the information on placement and best available cluster can sometimes be a one-shot and sometimes desire re-evaluation when best availability changes. Implementations and consumers will probably want to be aware of those scenarios as it will influence the way their loops work.

I assume that "implementation" == producer here. I am not sure that the producer needs to be aware of how the schedulingDecision is being consumed. The consumer can consume it once or do periodic re-evaluation or even event driven rescheduling. It's up to the consumer but the content of the schedulingDecision CR is immutable.

mikeshng commented on keps/sig-multicluster/5313-placement-decision-api/README.md:
```diff
@@ -0,0 +1,625 @@
+* Ensuring slice size <= 100 and required labels exists.
+
+* Integration tests: fake scheduler writes decisions and fake consumer verifies watches.
+  Scale test with large number of clusters and placements.
```
Right. Removed as suggested.

mikeshng commented:
I've updated the KEP to address the concerns raised during the previous [meeting](https://www.youtube.com/watch?v=WnPGVXj-QFY). Added clear explanations of the integration reduction benefit and what this API does and does not standardize. Expanded user stories with concrete label examples and reorganized the KEP structure for better readability.

PTAL Thanks!
